examples/Cassie/osc_jump/jumping_event_based_fsm.cc

```diff
@@ -0,0 +1,190 @@
+  }
+  this->DeclareVectorOutputPort(BasicVector<double>(1),
+                                &JumpingEventFsm::CalcFiniteState);
+  DeclarePerStepDiscreteUpdateEvent(&JumpingEventFsm::DiscreteVariableUpdate);
+
+  BasicVector<double> init_prev_time = BasicVector<double>(VectorXd::Zero(1));
+  BasicVector<double> init_state_trigger_time =
+      BasicVector<double>(VectorXd::Zero(1));
+  BasicVector<double> init_fsm_state = BasicVector<double>(VectorXd::Zero(1));
+  init_state_trigger_time.get_mutable_value()(0) = -1.0;
+  init_fsm_state.get_mutable_value()(0) = init_state_;
+
+  prev_time_idx_ = this->DeclareDiscreteState(init_prev_time);
+  guard_trigger_time_idx_ = this->DeclareDiscreteState(init_state_trigger_time);
+  fsm_idx_ = this->DeclareDiscreteState(init_fsm_state);
+  transition_flag_idx_ = this->DeclareDiscreteState(1);
+}
+
+EventStatus JumpingEventFsm::DiscreteVariableUpdate(
+    const Context<double>& context,
+    DiscreteValues<double>* discrete_state) const {
+  // Get inputs to the leaf system
+  const auto state_feedback =
+      this->template EvalVectorInput<OutputVector>(context, state_port_);
+  const drake::AbstractValue* contact_info =
+      this->EvalAbstractInput(context, contact_port_);
+  // Get the discrete states
+  auto fsm_state =
+      discrete_state->get_mutable_vector(fsm_idx_).get_mutable_value();
+  auto prev_time =
+      discrete_state->get_mutable_vector(prev_time_idx_).get_mutable_value();
+  auto state_trigger_time =
+      discrete_state->get_mutable_vector(guard_trigger_time_idx_)
+          .get_mutable_value();
+  auto transition_flag =
+      discrete_state->get_mutable_vector(transition_flag_idx_)
+          .get_mutable_value();
+
+  int num_contacts = 0;
+  if (simulator_type_ == DRAKE) {
+    const auto& contact_info_msg =
+        contact_info->get_value<drake::lcmt_contact_results_for_viz>();
+    num_contacts = contact_info_msg.num_point_pair_contacts;
+  } else if (simulator_type_ == MUJOCO) {
+    // MuJoCo has "persistent" contact so we have to check contact forces
+    // instead of just a boolean value of on/off
+    const auto& contact_info_msg =
+        contact_info->get_value<dairlib::lcmt_contact_mujoco>();
+    num_contacts = std::count_if(contact_info_msg.contact_forces.begin(),
+                                 contact_info_msg.contact_forces.end(),
+                                 [&](auto const& force) {
+                                   double threshold = 1e-6;
+                                   return std::abs(force) >= threshold;
+                                 });
+  }
+
+  double timestamp = state_feedback->get_timestamp();
+
+  // Simulator has restarted, reset FSM
+  if (timestamp < prev_time(0)) {
+    std::cout << "Simulator has restarted!" << std::endl;
+    fsm_state << init_state_;
+    prev_time(0) = timestamp;
+    transition_flag(0) = false;
+  }
+
+  // To test delayed switching times, there is an "intermediate" state
```

Comment:
The intermediate states are to test delayed controller switching times, maybe I should take it out of this commit and keep the delays in my experimental branch?

Reply:
Is this purely only for testing, or it can actually be used for delayed landing? 
Also I think if it's going to be used on the hardware later, maybe it can stay and we should add a TODO to delete this after hardware testing

Reply:
Oh good point. The contact based state transitions can only work in simulation though the time based fsm should be fine for hardware testing. Either way, the transition delays should be removed once the hardware testing is done. I'll add a TODO